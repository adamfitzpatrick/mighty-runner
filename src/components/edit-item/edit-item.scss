@import '../../assets/colors';

.editItem {
  display: block;
  border: 1px solid $color-white;
  border-radius: 0.5rem;
  margin: 0.25rem;
  padding: 0.5rem;
}
