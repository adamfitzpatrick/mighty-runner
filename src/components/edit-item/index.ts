export { default } from './edit-item'
export * from './edit-item'
