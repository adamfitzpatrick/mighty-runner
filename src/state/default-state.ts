import { Character, PersonalData, CharacterIdentifier, Attributes, Effect, GearItem, Pic } from '@models'

export interface AppState {
  characters: Character[] | null
  activeCharacter: CharacterIdentifier | null
  pic: Pic | null
  personalData: PersonalData | null
  attributes: Attributes | null
  gear: GearItem[] | null
  effects: Effect[] | null
  apiError: boolean
}

export const defaultState: AppState = {
  characters: null,
  pic: null,
  activeCharacter: null,
  personalData: null,
  attributes: null,
  gear: null,
  effects: null,
  apiError: false
}

export default defaultState
