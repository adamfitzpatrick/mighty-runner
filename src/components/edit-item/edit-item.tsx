import * as React from 'react'

import * as styles from './edit-item.scss'
import { Stat, Effect } from '@models'
import Button from '@components/button'
import { EffectsById } from '@containers/effects'

export interface EditItemRenderProp<T extends Stat> {
  (item: T, changeHandler: (item: T) => void): JSX.Element
}

interface Props<T extends Stat> {
  item: T
  render: EditItemRenderProp<T>
  changeHandler: (item: T) => void
  done: () => void,
  cancel?: () => void
}

export default function EditItem<T extends Stat> (
  { item, render, changeHandler, done, cancel }: Props<T>
) {
  const effects = item.effects

  const handleAddedEffect = (effect: Effect) => {
    const updated = Object.assign({}, item)
    updated.effects = (updated.effects || []).concat(effect.id)
    changeHandler(updated)
  }

  return (
    <div data-testid={`${item!.name}.edit-item.component`} className={styles.editItem}>
      {render(item!, changeHandler)}
      <EffectsById ids={ effects || [] } edit onAdd={handleAddedEffect} />
      <Button label='Done' onClick={ done } />
      { cancel ? <Button label='Cancel' onClick={ cancel } /> : null }
    </div>
  )
}
