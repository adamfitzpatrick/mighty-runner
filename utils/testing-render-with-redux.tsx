import * as React from 'react'
import {
  createStore,
  combineReducers,
  AnyAction,
  Store,
  Dispatch,
  applyMiddleware
} from 'redux'
import { Provider } from 'react-redux'
import { render, RenderResult } from '@testing-library/react'
import * as Models from '@models'
import { AppState } from '../src/state/default-state'

export type RenderWithRedux =
  RenderResult &
  {
    store: Store<AppState, AnyAction>,
    dispatchSpy: jest.Mock<Dispatch<AnyAction>>
  }

const reducer = combineReducers({
  characters: (state: Models.Character[] | null = null) => state,
  activeCharacter: (state: Models.CharacterIdentifier | null = null) => state,
  personalData: (state: Models.PersonalData | null = null) => state,
  attributes: (state: Models.Attributes | null = null) => state,
  gear: (state: Models.GearItem[] | null = null) => state,
  effects: (state: Models.Effect[] | null = null) => state
})

const getSpyMiddleware = (dispatchSpy: jest.Mock<Dispatch<AnyAction>>) => () => {
  return (next: Dispatch<AnyAction>) => (action: AnyAction) => {
    dispatchSpy(action)
    next(action)
  }
}

export function renderWithRedux (
  ui: JSX.Element,
  initialState: AppState
): RenderWithRedux {
  const dispatchSpy: jest.Mock<Dispatch<AnyAction>> = jest.fn()
  const spyMiddleware = getSpyMiddleware(dispatchSpy)
  const store = createStore(reducer, initialState, applyMiddleware(spyMiddleware))

  return {
    ...render(<Provider store={store}>{ ui }</Provider>),
    store,
    dispatchSpy
  }
}
