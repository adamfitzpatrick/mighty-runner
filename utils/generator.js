const fs = require('fs')
const path = require('path')
const yargs = require('yargs')

if (!yargs.argv.container && !yargs.argv.component) {
  throw new Error('no component or container name provided')
}

const container = yargs.argv.container && yargs.argv.container.replace(/^[a-z]/, a => a.toUpperCase())
const component = yargs.argv.component && yargs.argv.component.replace(/^[a-z]/, a => a.toUpperCase())

let thing = yargs.argv.component
if (container) {
  thing = container
}
const lower = thing.replace(/^[A-Z]/, a => a.toLowerCase()).replace(/[A-Z]/g, a => `-${a.toLowerCase()}`)
thing = thing.replace(/^[a-z]/, a => a.toUpperCase()).replace(/-([A-Za-z0-9])/g, a => a.slice(1).toUpperCase())
scssClass = thing.replace(/^[A-Z]/, a => a.toLowerCase())

const thingTs = `import * as React from 'react'

import * as styles from './${lower}.scss'

interface Props {}

export default function ${thing} (props: Props) {
  return <div>${thing}</div>
}
`

const thingTestTs = `import * as React from 'react'
import { cleanup, render } from '@testing-library/react'
import { renderWithRedux, RenderWithRedux } from '../../../utils/testing-render-with-redux'
import ${thing} from '.'

describe('${thing} ${ container ? 'container' : 'component'}', () => {
  test('should do nothing', () => {
    expect(true).toEqual(true)
  })
})
`

const indexTs = `export { default } from './${lower}'
`

const scss = `.${scssClass} {
  display: block;
}
`

const scssD = `export const ${scssClass}: string
`

let dest = path.resolve(process.cwd(), 'src', 'components', lower)
if (container) {
  dest = path.resolve(process.cwd(), 'src', 'containers', lower)
}

fs.mkdirSync(dest)
fs.writeFileSync(path.resolve(dest, `${lower}.tsx`), thingTs)
fs.writeFileSync(path.resolve(dest, `${lower}.test.tsx`), thingTestTs)
fs.writeFileSync(path.resolve(dest, 'index.ts'), indexTs)
fs.writeFileSync(path.resolve(dest, `${lower}.scss`), scss)
fs.writeFileSync(path.resolve(dest, `${lower}.scss.d.ts`), scssD)
