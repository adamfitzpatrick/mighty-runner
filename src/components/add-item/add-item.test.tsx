import * as React from 'react'
import * as Models from '@models'
import { RenderResult } from '@testing-library/react'
import AddItem from '.'
import { renderWithRedux } from '../../../utils/testing-render-with-redux'
import { AppState } from '@state/default-state'
import { EditItemRenderProp } from '@components/edit-item'

describe('AddItem component', () => {
  let appState: AppState
  let item: Models.Stat
  let changeSpy: jest.Mock<Models.Stat>
  let renderFunc: EditItemRenderProp<Models.Stat>
  let passedObjects: { item?: Models.Stat, changeHandler?: (item: Models.Stat) => void }
  let sut: RenderResult

  beforeEach(() => {
    appState = {} as AppState
    item = {
      id: '1',
      name: 'n',
      shortName: 'n',
      description: 'd',
      value: { thing: 1 },
      asEffectTarget: [ 'a', 'b' ]
    }
    changeSpy = jest.fn()
    renderFunc = (item: Models.Stat, changeHandler: (item: Models.Stat) => void): JSX.Element => {
      passedObjects = { item, changeHandler }
      return <div>ok</div>
    }
    sut = renderWithRedux(<AddItem
      item={ item }
      render={ renderFunc }
      changeHandler={ changeSpy }
      done={ () => { /* no-op */ } }
      cancel={ () => { /* no-op */ } }
    />, appState)
  })

  test('should render correctly', () => {
    expect(sut.container.innerHTML).toMatchSnapshot()
  })

  test('should pass the item and changeHandler into the rendered component', () => {
    expect(passedObjects.item).toBe(item)
    expect(passedObjects.changeHandler).toBe(changeSpy)
  })
})
