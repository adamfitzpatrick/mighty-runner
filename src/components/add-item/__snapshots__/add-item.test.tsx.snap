// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddItem component should render correctly 1`] = `"<div data-testid=\\"n.edit-item.component\\"><div>ok</div><button data-testid=\\"Done.button.component\\">Done</button><button data-testid=\\"Cancel.button.component\\">Cancel</button></div>"`;
